### ISSUE TITLE
sd-webui-tagcomplete fails to load in Auto

### ISSUE THREAD

**Issue by Rizz226** [AbdBarho/stable-diffusion-webui-docker#216]
**Has this issue been opened before?**

- [x] It is not in the [FAQ](https://github.com/AbdBarho/stable-diffusion-webui-docker/wiki/FAQ), I checked.
- [X] It is not in the [issues](https://github.com/AbdBarho/stable-diffusion-webui-docker/issues?q=), I searched.

**Describe the bug**
tag-autocomplete (and other extensions maybe?) fail to start due to missing files in ../repositories/stable-diffusion.

**Which UI**
auto
hlky or auto or auto-cpu or lstein?

error message:
Attaching to webui-docker-auto-1
webui-docker-auto-1  | Mounted .cache
webui-docker-auto-1  | Mounted LDSR
webui-docker-auto-1  | Mounted Hypernetworks
webui-docker-auto-1  | Mounted VAE
webui-docker-auto-1  | Mounted GFPGAN
webui-docker-auto-1  | Mounted RealESRGAN
webui-docker-auto-1  | Mounted ScuNET
webui-docker-auto-1  | Mounted .cache
webui-docker-auto-1  | Mounted StableDiffusion
webui-docker-auto-1  | Mounted embeddings
webui-docker-auto-1  | Mounted ESRGAN
webui-docker-auto-1  | Mounted config.json
webui-docker-auto-1  | Mounted Deepdanbooru
webui-docker-auto-1  | Mounted SwinIR
webui-docker-auto-1  | Mounted ui-config.json
webui-docker-auto-1  | Mounted BSRGAN
webui-docker-auto-1  | Mounted Codeformer
webui-docker-auto-1  | Mounted extensions
webui-docker-auto-1  | + python3 -u ../../webui.py --listen --port 7860 --ckpt-dir /stable-diffusion-webui/models/Stable-diffusion --enable-insecure-extension-access --allow-code --medvram --xformers --listen
webui-docker-auto-1  | Removing empty folder: /stable-diffusion-webui/models/BSRGAN
webui-docker-auto-1  | Error loading script: tag_autocomplete_helper.py
webui-docker-auto-1  | Traceback (most recent call last):
webui-docker-auto-1  |   File "/stable-diffusion-webui/modules/scripts.py", line 170, in load_scripts
webui-docker-auto-1  |     exec(compiled, module.__dict__)
webui-docker-auto-1  |   File "/stable-diffusion-webui/extensions/a1111-sd-webui-tagcomplete/scripts/tag_autocomplete_helper.py", line 77, in <module>
webui-docker-auto-1  |     write_tag_base_path()
webui-docker-auto-1  |   File "/stable-diffusion-webui/extensions/a1111-sd-webui-tagcomplete/scripts/tag_autocomplete_helper.py", line 63, in write_tag_base_path
webui-docker-auto-1  |     f.write(TAGS_PATH.relative_to(FILE_DIR).as_posix())
webui-docker-auto-1  |   File "/usr/local/lib/python3.10/pathlib.py", line 818, in relative_to
webui-docker-auto-1  |     raise ValueError("{!r} is not in the subpath of {!r}"
webui-docker-auto-1  | ValueError: '/stable-diffusion-webui/extensions/a1111-sd-webui-tagcomplete/tags' is not in the subpath of '/stable-diffusion-webui/repositories/stable-diffusion' OR one path is relative and the other is absolute.

**Hardware / Software**
 - OS: Ubuntu
 - OS version: 20.04
 - Docker Version: 
 - Docker compose version: 
 - Repo version: Master, 11/11/2022
 - RAM: 64gb
 - GPU/VRAM: 3080 10gb

**Steps to Reproduce**
Follow autocomplete's install instructions.

**Additional context**
Tried the copy action in pull #215, but it did not work. 

I totally get that this might be an issue with the extension and out of scope for you, but figured it was worth bringing up because it seems to work fine on a bare metal install.

**Comment by AbdBarho** [comment]
@Rizz226 the extensions folder is mounted, so copying should not be necessary, I am not sure what the problem is or what the PR solves.

**Comment by AbdBarho** [comment]
@Rizz226 you are correct extensions is out of reach of this repo.

That being said, you can try to hack around it in your `starup.sh`, you can maybe `sed` replace the line or run your own python script that does the same, or even manually change it.

**Comment by DominikDoom** [comment]
To give some insight into this as the dev of tag-autocomplete, at least from the error message the issue seems to be that the working directory is different from the execution directory of the webui, leading to failing path checks.

My extension needs to know the location of some files relative to the root since I need to access them in the client-side JavaScript, which only allows direct requests for safety reasons. Meaning I can't scan folders like in python but prepare path listings at startup instead with a python helper script.
This is similar to issues some colab users have run into in the past.
If you can't change the working dir, or if it *needs* to be different from the root, it might work to change line 8 of `scripts/tag_autocomplete_helper.py`:
```py
# Webui root path
FILE_DIR = Path().absolute()
```
Currently this seems to point to the working dir, so `/stable-diffusion-webui/repositories/stable-diffusion`.
But the file path reported by the webui's `scripts.basedir()` property is `/stable-diffusion-webui/extensions/a1111-sd-webui-tagcomplete`, so for the extension to work, FILE_DIR should just be `/stable-diffusion-webui`.
I'm not sure if there is a better way to find out the actual root in this helper script, but as a docker-specific solution `sed`-ing the line like @AbdBarho said or forking it for this purpose might work.

**Comment by AbdBarho** [comment]
@DominikDoom thank you for the insights! this makes the problem a lot clearer.

I remember from the early beginning that the webui should have cwd of the original compviz repo under `repositories` folder, but I think I have seen some code that indicates that this is not necessary anymore.

Anyway, I created a pull request with the adjusted cwd: https://github.com/AbdBarho/stable-diffusion-webui-docker/pull/230

However, now I get a new error, which seems to be related to gradio:

```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/uvicorn/protocols/http/h11_impl.py", line 407, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/usr/local/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/fastapi/applications.py", line 270, in __call__
    await super().__call__(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 24, in __call__
    await responder(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 43, in __call__
    await self.app(scope, receive, self.send_with_gzip)
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 75, in __call__
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 64, in __call__
    await self.app(scope, receive, sender)
  File "/usr/local/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/usr/local/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 680, in __call__
    await route.handle(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 275, in handle
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 65, in app
    response = await func(request)
  File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 235, in app
    raw_response = await run_endpoint_function(
  File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 163, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
  File "/usr/local/lib/python3.10/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
  File "/usr/local/lib/python3.10/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/usr/local/lib/python3.10/site-packages/gradio/routes.py", line 243, in file_deprecated
    return file(path)
  File "/usr/local/lib/python3.10/site-packages/gradio/routes.py", line 236, in file
    raise ValueError(
ValueError: File cannot be fetched: extensions/a1111-sd-webui-tagcomplete/tags/config.json, perhaps because it is not in any of {'/tmp/tmp3c5rskc8', '/tmp/tmp1ksgrvht', '/tmp/tmpd7fm3yh1', '/tmp/tmp2d41600c', '/tmp/tmpwk1ly7kl', '/tmp/tmpwmd7olco', '/tmp/tmpq4a3wwnp', '/tmp/tmpvwhep26n', '/tmp/tmpov_io9p2', '/tmp/tmp_bmxp55n', '/tmp/tmpd9auct1f', '/tmp/tmpgmuz3kp5', '/tmp/tmpq30u5g76', '/tmp/tmpyuxv8hep', '/tmp/tmp85yj4vup'}
```

I know that recently gradio introduced a "fix" for the performance degradation in their UI after generating a lot of images (due to the fact that all images were transferred as base64 string), and the fix was to copy the generated images over to a temp directory and serve them from there, and the temp paths look like the above.

The config file seems to be loaded by an http request:
https://github.com/DominikDoom/a1111-sd-webui-tagcomplete/blob/fc4484ddc6fbcd16b9f1ae311243c9696308675a/javascript/tagAutocomplete.js#L695
but the folder is not served by gradio, which is why it is complaining.

Is there a way to tell gradio to also serve the cwd? 
maybe downgrading gradio would help?


BTW, `XMLHttpRequest` is deprecated, you can just
```js
acConfig = await fetch(`file/${tagBasePath}/config.json?${new Date().getTime()}`).then(r => r.json());
```

**Comment by DominikDoom** [comment]
Yeah, that seems to be a bigger issue.
Gradio serves all files in the app root and subfolders by default. All other local files aren't accessible for safety. I think there are so many different IDs in the message because Gradio seemingly keeps separate temp directories to serve each block: https://github.com/gradio-app/gradio/blob/831ae1405f0f2cddf8d508775731eef39c5a09df/gradio/routes.py#L253-L256.
But I'm not really sure why Gradio doesn't serve these files in your case. Maybe Docker prevents Gradio from filling those temp dirs correctly?

> BTW, XMLHttpRequest is deprecated, you can just use fetch

Good call, the reason it's still on XMLHttpRequest is that it used synchronous mode at the beginning, which fetch doesn't support. But I already refactored it to use async a while ago, so it's true that I can switch to fetch now.

**Comment by AbdBarho** [comment]
@DominikDoom in the gradio link you sent: https://github.com/gradio-app/gradio/blob/831ae1405f0f2cddf8d508775731eef39c5a09df/gradio/routes.py#L230 The path seems to be in the form of `file=...`, but in the fetch call https://github.com/DominikDoom/a1111-sd-webui-tagcomplete/blob/fc4484ddc6fbcd16b9f1ae311243c9696308675a/javascript/tagAutocomplete.js#L695, it is `file/${tagBasePath}/config.json`, do you think this makes a difference, or is it redirected anyway?

**Comment by DominikDoom** [comment]
Good to know that that url form seems to be deprecated in favor of the `file=...` one. I tried just now to make sure, on my local installation both work as expected.

**Comment by AbdBarho** [comment]
@DominikDoom it seems that I found the problem

TLDR: it is related to the mounts / symlinks in the docker container.

For debugging, I added 
```py
print(Path(path).resolve())
```
right before this line https://github.com/gradio-app/gradio/blob/831ae1405f0f2cddf8d508775731eef39c5a09df/gradio/routes.py#L234
and got:
```
/data/config/auto/extensions/a1111-sd-webui-tagcomplete/tags/config.json
```
instead of the expected 
```
/stable-diffusion-webui/extensions/a1111-sd-webui-tagcomplete/tags/config.json
```

`/data/config/auto` is the folder in which the extensions are saved and later symlinked into `/stable-diffusion-webui/extensions`

https://github.com/AbdBarho/stable-diffusion-webui-docker/blob/8623c737411e79fdeb80be713a75b2cdb87736f3/services/AUTOMATIC1111/entrypoint.sh#L38

**Comment by DominikDoom** [comment]
So is the problem that Gradio uses the original path when it should be the symlinked path? That seems tricky to solve.

> I am unsure of what would be the best way to handle this problem, do you have any suggestions? 
A hard copy of the extensions would work, but they will be lost once the container is deleted.

I'm not familiar with docker so I don't think I will be of much help in that regard. The only thing I found regarding a similar issue is using `readlink -f` to get the actual file path of a symlink, but I'm not sure how or if we could pass that so gradio searches the right spot.
